pkg/virt-controller/watch/vm/vm.go

```diff
@@ -2056,28 +2054,27 @@ func hasStopRequestForVMI(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineI
 }
 
 // no special meaning, randomly generated on my box.
-// TODO: do we want to use another constants? see examples in RFC4122
 const magicUUID = "6a1a24a1-4061-4607-8bf4-a3963d0c5895"
 
 var firmwareUUIDns = uuid.MustParse(magicUUID)
 
-// setStableUUID makes sure the VirtualMachineInstance being started has a 'stable' UUID.
-// The UUID is 'stable' if doesn't change across reboots.
-func setupStableFirmwareUUID(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) {
-
-	logger := log.Log.Object(vm)
-
-	if vmi.Spec.Domain.Firmware == nil {
-		vmi.Spec.Domain.Firmware = &virtv1.Firmware{}
+// syncFirmwareUUID makes sure VirtualMachine has a stable UUID persisted in its spec
+func syncFirmwareUUID(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) {
+	if vm.Spec.Template.Spec.Domain.Firmware == nil {
+		vm.Spec.Template.Spec.Domain.Firmware = &virtv1.Firmware{}
 	}
 
-	existingUUID := vmi.Spec.Domain.Firmware.UUID
+	existingUUID := vm.Spec.Template.Spec.Domain.Firmware.UUID
 	if existingUUID != "" {
-		logger.V(4).Infof("Using existing UUID '%s'", existingUUID)
 		return
 	}
 
-	vmi.Spec.Domain.Firmware.UUID = types.UID(uuid.NewSHA1(firmwareUUIDns, []byte(vmi.ObjectMeta.Name)).String())
+	// This code generates a non-uniqe identification. Two VMs with the
+	// same name but on different namespaces or on different clusters get
+	// the same ID. We have to keep this for a while because modifying this
+	// logic would modify the ID of existing VMs.
+	// https://github.com/kubevirt/kubevirt/issues/13156
+	vm.Spec.Template.Spec.Domain.Firmware.UUID = types.UID(uuid.NewSHA1(firmwareUUIDns, []byte(vmi.ObjectMeta.Name)).String())
```

Comment:
This part is problematic, as it's invalid to modify the vm here as it's the cached object from the informer's store, which will leave the object in an inconsistent state.

This code is eventually being called from here:
https://github.com/kubevirt/kubevirt/blob/96c9f1180eb8770d58a1f74eb70f0a569d074e8e/pkg/virt-controller/watch/vm/vm.go#L3280

a few lines later, the vm is being copied. Only the copy should be modified, leaving the cached object unchanged:
https://github.com/kubevirt/kubevirt/blob/96c9f1180eb8770d58a1f74eb70f0a569d074e8e/pkg/virt-controller/watch/vm/vm.go#L3294

So even if we follow this approach, we'd have to do it in the scope above. Let me hold the PR in the meantime.

/hold

Reply:
> An alternative to the approach in this PR is adding a firmware UUID field to the VM's status. The obvious con is that we'll be adding an API field which is probably not very important outside the scope of this bug.

I trust you that my implementation is bad, but 'status' is unlikely to help. VMs should have a persistent firmware UUID to be reused on the next time the VM is started.

Can you suggest another place where one can fill Firmware.UUID of existing VMs, e.g during virt-controller startup?

Reply:
> but 'status' is unlikely to help. VMs should have a persistent firmware UUID to be reused on the next time the VM is started.

Sorry, perhaps I'm missing something, but why wouldn't a status field solve this problem?
My idea is that whenever the VM would boot for the first time, and create a UUID, this UUID will be saved to the VM's status in a persistent manner. This way whenever the VM would boot, we would first try to see if the status UUID field is populated, and if it is we'd use this saved UUID.

Am I missing anything?

Reply:
I’d like to note that this is around the fourth time the topic of persistent firmware UUID is discussed. Each conversation has its own merits, but I particularly agree with  https://github.com/kubevirt/kubevirt/pull/12085#issuecomment-2275384536. This issue calls for a design proposal IMO.

Reply:
> > but 'status' is unlikely to help. VMs should have a persistent firmware UUID to be reused on the next time the VM is started.
> 
> Sorry, perhaps I'm missing something, but why wouldn't a status field solve this problem? My idea is that whenever the VM would boot for the first time, and create a UUID, this UUID will be saved to the VM's status in a persistent manner. This way whenever the VM would boot, we would first try to see if the status UUID field is populated, and if it is we'd use this saved UUID.
> 
> Am I missing anything?

I like this idea. It fits the same pattern we have with CurrentCPUTopology and CurentMemeory, as well as the pattern we were trying to push for the migrated volumes

Reply:
> However, the pro is that this approach uses the right separation between spec (the user's desired state) and status (the actual state). Ideally, the spec should contain only whatever the user cares about and leave everything else blank.

Patching the spec of a running VM can be tricky. When the `rolloutStrategy` is set to `liveUpdate`, patching the VM spec would mean that we will raise the restart required condition.
Most of the VM should be running after KubeVirt's upgrade, hence we will run into this issue, but not if we will store the UUID in the VM status.

Reply:
Thank you all for the discussion.
Let's please continue it on @dasionov's design proposal here: https://github.com/kubevirt/community/pull/347.

I've tried to list all of the different alternatives in https://github.com/kubevirt/community/pull/347#pullrequestreview-2420156264. Let me know if you have more alternatives in mind.